File: 2-add-two-numbers/add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    // Initialize a dummy head for the result linked list
    let dummyHead = new ListNode(0);
    let current = dummyHead;
    
    // Initialize carry value
    let carry = 0;
    
    // Traverse both linked lists
    while (l1 !== null || l2 !== null) {
        // Get values from the current nodes (or 0 if node is null)
        let x = l1 !== null ? l1.val : 0;
        let y = l2 !== null ? l2.val : 0;
        
        // Calculate sum and new carry
        let sum = x + y + carry;
        carry = Math.floor(sum / 10);
        
        // Create a new node with digit value (sum % 10)
        current.next = new ListNode(sum % 10);
        current = current.next;
        
        // Move to next nodes if available
        if (l1 !== null) l1 = l1.next;
        if (l2 !== null) l2 = l2.next;
    }
    
    // If there's a remaining carry, add a new node
    if (carry > 0) {
        current.next = new ListNode(carry);
    }
    
    // Return the result list (excluding the dummy head)
    return dummyHead.next;
};

File: 2-add-two-numbers/add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        // Create a dummy head node
        ListNode* dummyHead = new ListNode(0);
        ListNode* current = dummyHead;
        
        // Initialize carry
        int carry = 0;
        
        // Traverse both linked lists
        while (l1 != nullptr || l2 != nullptr) {
            // Get values from current nodes (or 0 if node is null)
            int x = (l1 != nullptr) ? l1->val : 0;
            int y = (l2 != nullptr) ? l2->val : 0;
            
            // Calculate sum and new carry
            int sum = x + y + carry;
            carry = sum / 10;  // Integer division in C++
            
            // Create new node with digit value (sum % 10)
            current->next = new ListNode(sum % 10);
            current = current->next;
            
            // Move to next nodes if available
            if (l1 != nullptr) l1 = l1->next;
            if (l2 != nullptr) l2 = l2->next;
        }
        
        // If there's a remaining carry, add a new node
        if (carry > 0) {
            current->next = new ListNode(carry);
        }
        
        // Store the result (excluding dummy head)
        ListNode* result = dummyHead->next;
        
        // Free the dummy head to avoid memory leak
        delete dummyHead;
        
        // Return the result list
        return result;
    }
};